File: crates/clang-importer-sys/cxx/lib/cxx/swift/LangOptions.cxx
#include "cxx/swift/LangOptions.hxx"

namespace cxx {
namespace swift {
namespace LangOptions {

std::unique_ptr<::swift::LangOptions> make() {
  return std::make_unique<::swift::LangOptions>(::swift::LangOptions());
}

std::unique_ptr<::llvm::Triple> Target(::swift::LangOptions const &This) {
  return std::make_unique<::llvm::Triple>(This.Target);
}

void SetTarget(::swift::LangOptions &This,
               std::unique_ptr<::llvm::Triple> target) {
  This.Target = *target;
}

} // namespace LangOptions
} // namespace swift
} // namespace cxx

File: crates/clang-importer-sys/rust/src/ffi/swift/lang_options.rs
#[cxx::bridge(namespace = "rust::swift")]
pub(crate) mod ffi {
    struct LangOptions {
        ptr: UniquePtr<CxxLangOptions>,
    }

    #[namespace = "swift"]
    extern "C++" {
        include!("swift/Basic/LangOptions.h");

        #[cxx_name = "LangOptions"]
        type CxxLangOptions;
    }

    #[namespace = "cxx::swift::LangOptions"]
    extern "C++" {
        include!("cxx/swift/LangOptions.hxx");

        unsafe fn make() -> UniquePtr<CxxLangOptions>;

        #[namespace = "llvm"]
        #[cxx_name = "Triple"]
        type CxxTriple = crate::ffi::llvm::triple::ffi::CxxTriple;

        unsafe fn Target(This: &CxxLangOptions) -> UniquePtr<CxxTriple>;

        unsafe fn SetTarget(This: Pin<&mut CxxLangOptions>, target: UniquePtr<CxxTriple>);
    }
}

use self::ffi::{CxxLangOptions, LangOptions};
use crate::llvm::Triple;
use cxx::UniquePtr;

impl From<UniquePtr<CxxLangOptions>> for LangOptions {
    #[inline]
    fn from(ptr: UniquePtr<CxxLangOptions>) -> Self {
        Self { ptr }
    }
}

impl LangOptions {
    #[inline]
    pub unsafe fn new() -> Self {
        let ptr = self::ffi::make();
        Self { ptr }
    }

    #[inline]
    pub unsafe fn target(&self) -> Triple {
        let this = &self.ptr;
        let ptr = self::ffi::Target(this);
        Triple { ptr }
    }

    #[inline]
    pub unsafe fn set_target(&mut self, target: Triple) {
        let this = self.ptr.pin_mut();
        self::ffi::SetTarget(this, target.ptr);
    }
}
